src/server/page_server.cpp

```diff
@@ -277,18 +364,27 @@ page_server::set_passive_tran_server_connection (cubcomm::channel &&chn)
 void
 page_server::disconnect_active_tran_server ()
 {
-  er_log_debug (ARG_FILE_LINE, "Page server disconnected from active transaction server with channel id: %s.\n",
-		m_active_tran_server_conn->get_channel_id ().c_str ());
-  m_active_tran_server_conn.reset (nullptr);
+  if (m_active_tran_server_conn != nullptr)
+    {
+      er_log_debug (ARG_FILE_LINE, "disconnect_active_tran_server:"
+		    " Disconnect active transaction server connection with channel id: %s.\n",
+		    m_active_tran_server_conn->get_channel_id ().c_str ());
+      m_active_tran_server_conn.reset (nullptr);
+    }
+  else
+    {
+      er_log_debug (ARG_FILE_LINE, "disconnect_active_tran_server: Active transaction server is not connected.\n");
+    }
 }
 
 void
-page_server::disconnect_tran_server (connection_handler *conn)
+page_server::disconnect_tran_server_async (connection_handler *conn)
 {
   assert (conn != nullptr);
   if (conn == m_active_tran_server_conn.get ())
     {
-      disconnect_active_tran_server ();
+      m_disconnect_handler.disconnect (std::move (m_active_tran_server_conn));
```

Comment:
Adding/removing m_active_tran_server_conn, m_passive_tran_server_conn should be protected.

Reply:
Indeed, this discussion came in the past, but we thought it is not yet "time" to consider it.

For `m_active_tran_server_conn`:
- adding: is only done once, so it does not need to be
- deleting it sync or async: indeed it needs to be
- using it (eg: in `push_request_to_active_tran_server`) is only done in one place, but usage is not synchronized with deletion
Same goes for each instance of `m_passive_tran_server_conn`.
I think it is better to take this as a separate issue - I've added http://jira.cubrid.org/browse/LETS-408 .

Reply:
I agree to handle it in a separate issue.